File: pointers/c/mutables.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Structure to demonstrate nested data
typedef struct {
    int scores[3];
    char name[50];
} User;

void print_array(int* arr, int size) {
    printf("[");
    for (int i = 0; i < size; i++) {
        printf("%d%s", arr[i], i < size - 1 ? ", " : "");
    }
    printf("]");
}

void print_state(const char* name1, int* ptr1, const char* name2, int* ptr2) {
    printf("\n%s points to address: %p", name1, (void*)ptr1);
    printf("\n%s points to address: %p", name2, (void*)ptr2);
    printf("\nSame address? %s", ptr1 == ptr2 ? "Yes" : "No");
    if (ptr1 && ptr2) {
        printf("\nValues: %s = %d, %s = %d", name1, *ptr1, name2, *ptr2);
    }
    printf("\n");
}

void demonstrate_basic_pointers() {
    printf("\n=== Basic Pointer Behavior ===\n");
    
    int x = 42;
    int* ptr1 = &x;
    int* ptr2 = ptr1;    // Both pointers point to x
    
    printf("\nInitial state:");
    print_state("ptr1", ptr1, "ptr2", ptr2);
    
    // Modify through ptr1
    *ptr1 = 100;
    printf("\nAfter modifying through ptr1:");
    print_state("ptr1", ptr1, "ptr2", ptr2);
    
    // Create new integer and point ptr2 to it
    int y = 200;
    ptr2 = &y;
    printf("\nAfter pointing ptr2 to new value:");
    print_state("ptr1", ptr1, "ptr2", ptr2);
}

void demonstrate_array_pointers() {
    printf("\n=== Array Pointer Behavior ===\n");
    
    // Create two arrays
    int arr1[] = {1, 2, 3};
    int arr2[] = {1, 2, 3};
    
    // Create pointers to arrays
    int* ptr1 = arr1;
    int* ptr2 = arr2;
    
    printf("\nInitial arrays:");
    printf("\narr1: ");
    print_array(arr1, 3);
    printf("\narr2: ");
    print_array(arr2, 3);
    printf("\n");
    
    // Show that arrays are at different locations
    printf("\nArray addresses:");
    print_state("arr1", ptr1, "arr2", ptr2);
    
    // Modify through pointer
    ptr1[1] = 99;
    printf("\nAfter modifying arr1 through pointer:");
    printf("\narr1: ");
    print_array(arr1, 3);
    printf("\narr2: ");
    print_array(arr2, 3);
    printf("\n");
}

void demonstrate_nested_structures() {
    printf("\n=== Nested Structure Behavior ===\n");
    
    // Create and initialize two users
    User* user1 = malloc(sizeof(User));
    User* user2 = malloc(sizeof(User));
    
    // Initialize user1
    strcpy(user1->name, "Alice");
    user1->scores[0] = 85;
    user1->scores[1] = 90;
    user1->scores[2] = 95;
    
    // Create a "shallow copy" by copying memory
    *user2 = *user1;
    strcpy(user2->name, "Bob");  // Give second user different name
    
    printf("\nAfter shallow copy:");
    printf("\nUser1: %s, scores: ", user1->name);
    print_array(user1->scores, 3);
    printf("\nUser2: %s, scores: ", user2->name);
    print_array(user2->scores, 3);
    
    // Modify user2's scores
    user2->scores[1] = 100;
    
    printf("\n\nAfter modifying user2's scores:");
    printf("\nUser1: %s, scores: ", user1->name);
    print_array(user1->scores, 3);
    printf("\nUser2: %s, scores: ", user2->name);
    print_array(user2->scores, 3);
    
    free(user1);
    free(user2);
}

void demonstrate_function_pointers() {
    printf("\n=== Function Parameter Behavior ===\n");
    
    int x = 42;
    int* ptr = &x;
    
    printf("\nBefore function call:");
    printf("\nx = %d", x);
    printf("\nptr points to: %p", (void*)ptr);
    
    // Pass pointer to function
    int y = 100;
    ptr = &y;
    
    printf("\n\nAfter changing pointer:");
    printf("\nx = %d", x);
    printf("\ny = %d", y);
    printf("\nptr now points to: %p", (void*)ptr);
    printf("\nptr value: %d\n", *ptr);
}

int main() {
    demonstrate_basic_pointers();
    demonstrate_array_pointers();
    demonstrate_nested_structures();
    demonstrate_function_pointers();
    return 0;
}

File: pointers/python/mutables.py
def print_state(name1, val1, name2, val2):
    """Print the current state of two variables and their memory locations."""
    print(f"\n{name1} = {repr(val1)}")
    print(f"{name2} = {repr(val2)}")
    print(f"Address of {name1}: {id(val1)}")
    print(f"Address of {name2}: {id(val2)}")
    print(f"Same object? {val1 is val2}")

def demonstrate_default_args_pitfall():
    """Demonstrate the dangerous default mutable argument pitfall."""
    print("\n=== Mutable Default Argument Pitfall ===")
    
    def add_to_list(item, target_list=[]):  # DANGEROUS: mutable default argument
        target_list.append(item)
        return target_list
    
    print("First call: add_to_list(1)")
    result1 = add_to_list(1)
    print(f"Result: {result1}")
    
    print("\nSecond call: add_to_list(2)")
    result2 = add_to_list(2)
    print(f"Result: {result2}")  # Unexpectedly contains [1, 2]!
    
    print("\nProof that both calls used the same list:")
    print(f"First result points to:  {id(result1)}")
    print(f"Second result points to: {id(result2)}")
    
    print("\nCorrect version:")
    def add_to_list_fixed(item, target_list=None):
        if target_list is None:
            target_list = []
        target_list.append(item)
        return target_list
    
    print("Using fixed version:")
    result3 = add_to_list_fixed(1)
    result4 = add_to_list_fixed(2)
    print(f"First call result:  {result3}")
    print(f"Second call result: {result4}")

def demonstrate_shallow_copy_pitfall():
    """Demonstrate issues with shallow copying of nested structures."""
    print("\n=== Shallow Copy Pitfall ===")
    
    # Create a nested structure
    original = {
        'name': 'User1',
        'scores': [85, 90, 95],
        'metadata': {'created': '2024-01-19'}
    }
    
    # Make a shallow copy
    import copy
    shallow_copy = copy.copy(original)
    
    print("Original:", original)
    print("Shallow copy:", shallow_copy)
    print("\nModifying nested list in shallow copy...")
    shallow_copy['scores'].append(100)
    
    print("\nUnexpected result - both dictionaries affected:")
    print("Original:", original)
    print("Shallow copy:", shallow_copy)
    
    # Show the correct way using deep copy
    deep_copy = copy.deepcopy(original)
    print("\nUsing deep copy instead...")
    deep_copy['scores'].append(105)
    print("Original:", original)
    print("Deep copy:", deep_copy)

def demonstrate_list_reference_pitfall():
    """Demonstrate issues with list references in loops."""
    print("\n=== List Reference in Loops Pitfall ===")
    
    # Create a list of lists - WRONG WAY
    wrong_matrix = [[0] * 3] * 3
    print("Initial matrix:", wrong_matrix)
    
    print("\nTrying to modify one element...")
    wrong_matrix[0][1] = 1
    print("Unexpected result - all rows modified:", wrong_matrix)
    
    # Correct way
    correct_matrix = [[0 for _ in range(3)] for _ in range(3)]
    print("\nCorrect matrix creation:")
    print("Initial state:", correct_matrix)
    correct_matrix[0][1] = 1
    print("After modification:", correct_matrix)

def demonstrate_dict_reference_pitfall():
    """Demonstrate issues with dictionary references in class attributes."""
    print("\n=== Dictionary Reference in Class Pitfall ===")
    
    class UserWrong:
        # DANGEROUS: Class-level mutable attribute
        settings = {'theme': 'light', 'notifications': True}
        
        def __init__(self, name):
            self.name = name
    
    # Create two users
    user1 = UserWrong('Alice')
    user2 = UserWrong('Bob')
    
    print("Initial settings:")
    print(f"User1 settings: {user1.settings}")
    print(f"User2 settings: {user2.settings}")
    
    print("\nModifying User1's settings...")
    user1.settings['theme'] = 'dark'
    
    print("\nUnexpected result - both users affected:")
    print(f"User1 settings: {user1.settings}")
    print(f"User2 settings: {user2.settings}")
    
    # Correct implementation
    class UserCorrect:
        def __init__(self, name):
            self.name = name
            # Instance-level dictionary
            self.settings = {'theme': 'light', 'notifications': True}
    
    print("\nUsing correct implementation:")
    user3 = UserCorrect('Charlie')
    user4 = UserCorrect('David')
    user3.settings['theme'] = 'dark'
    
    print(f"User3 settings: {user3.settings}")
    print(f"User4 settings: {user4.settings}")

def main():
    demonstrate_default_args_pitfall()
    demonstrate_shallow_copy_pitfall()
    demonstrate_list_reference_pitfall()
    demonstrate_dict_reference_pitfall()

if __name__ == "__main__":
    main()
